File: solutions/java/game-of-life/1/src/main/java/GameOfLife.java
class GameOfLife {
    public int[][] tick(int[][] matrix) {
        if(matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return new int[][]{};
        }
        int[][] new_matrix = new int[matrix.length][matrix[0].length];

        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                int current = matrix[i][j];
                int neighbors = count_closes(i, j, matrix);
                if (current == 0 && neighbors == 3) {
                    new_matrix[i][j] = 1;
                } else if (current == 1 && (neighbors < 2 || neighbors > 3)) {
                    new_matrix[i][j] = 0;
                } else {
                    new_matrix[i][j] = current;
                }
            }
        }
        return new_matrix;
    }

    private int count_closes(int x, int y, int[][] matrix) {
        int HEIGHT = matrix.length;
        int WIDTH = matrix[0].length;
        int count = 0;

        for (int i = -1; i <= 1; i++) {
            for (int j = -1; j <= 1; j++) {
                int row = x + i;
                int col = y + j;
                if (i == 0 && j == 0) continue;
                if (row >= 0 && row < HEIGHT && col >= 0 && col < WIDTH) {
                    count += matrix[row][col];
                }
            }
        }
        return count;
    }
}

File: solutions/dart/game-of-life/1/lib/game_of_life.dart
class GameOfLife {
  List<List<int>> cells = [];

  GameOfLife(this.cells);

  void tick() {
    if (cells.isEmpty) return;

    final rows = cells.length;
    final cols = cells[0].length;

    final next = List.generate(
      rows,
          (_) => List.filled(cols, 0),
    );

    for (var row = 0; row < rows; row++) {
      for (var col = 0; col < cols; col++) {
        final liveNeighbors = _countLiveNeighbors(row, col);
        final isAlive = cells[row][col] == 1;

        if (isAlive && (liveNeighbors == 2 || liveNeighbors == 3)) {
          next[row][col] = 1;
        } else if (!isAlive && liveNeighbors == 3) {
          next[row][col] = 1;
        }
      }
    }

    cells = next;
  }

  int _countLiveNeighbors(int row, int col) {
    var count = 0;

    for (var rowOffset = -1; rowOffset <= 1; rowOffset++) {
      for (var colOffset = -1; colOffset <= 1; colOffset++) {
        if (rowOffset == 0 && colOffset == 0) continue;

        final neighborRow = row + rowOffset;
        final neighborCol = col + colOffset;

        if (neighborRow < 0 ||
            neighborRow >= cells.length ||
            neighborCol < 0 ||
            neighborCol >= cells[neighborRow].length) {
          continue;
        }

        count += cells[neighborRow][neighborCol];
      }
    }

    return count;
  }

  List<List<int>> matrix() {
    return cells;
  }
}
